### PR TITLE
Add dynamic configuration features and test cases

### PR DESCRIPTION
**Pull request by JiaBaoxi**

### CODE REVIEW

File: roscpp_azure_iothub/README.md [hunk 1]
```diff
@@ -91,6 +92,50 @@ And add the `ros_relays` block to the device twin which you are about to connect
...
+Similar as the telemetry reporting, we are using the device twin `desired` properties as a channel to ask our ROS node what dynamic parameters to reconfigure.
+
+Here is a JSON example to reconfigure the parameters of `/dynamic_tutorials_node` with the new value:
+```
+{
+    "deviceId": "devA",
+    ...
+    "properties": {
+        "desired": {
+            "ros_dynamic_configurations": {
```
Comment (ooeygui) [inline roscpp_azure_iothub/README.md:106]: It looks like you are using map semantics for an array - why not use an array?

**Review by ooeygui — APPROVED** [review]

**Review by ooeygui — APPROVED** [review]

File: dynamic_tutorials/package.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>dynamic_tutorials</name>
+  <version>1.0.0</version>
+  <description>The dynamic_tutorials package</description>
+
+  <maintainer email="admin@contoso.com">admin</maintainer>
```
Comment (seanyen) [inline dynamic_tutorials/package.xml]: you can use your email alias here. :)

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp [hunk 3]
```diff
@@ -333,38 +453,28 @@ static int InitializeAzureIoTHub(ROS_Azure_IoT_Hub* iotHub)
     (void)IoTHubDeviceClient_SetConnectionStatusCallback(iotHub->deviceHandle, connection_status_callback, NULL);
     (void)IoTHubDeviceClient_SetDeviceTwinCallback(iotHub->deviceHandle, deviceTwinCallback, iotHub);  //Device Twin callback requires context to send message and device Handle
 
-    return 0;
+    return true;
 }
 
+static void DeinitializeAzureIoTHub(ROS_Azure_IoT_Hub* iotHub)
+{
+    ROS_INFO("Deinitializing IoTHub Device client\r\n");
```
Comment (seanyen) [inline roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: nit: I think it is okay to remove `\r\n` from ROS logging API.

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp [hunk 4]
```diff
@@ -274,55 +282,167 @@ static void subscribeTopic(const char* topicName, ROS_Azure_IoT_Hub* iotHub)
...
     }
 
-    std::unique_lock<std::mutex> lck(iotHub->mtx);
-    iotHub->signal = 1;
-    iotHub->cv.notify_all();
+    if (!g_is_startup)
+    {
+        // Note: We ignore the existing ros_dynamic_configurations settings during ros_azure_iothub_cpp_node starts up.
+        // Reason: Suppose the other nodes are working with the expected dynamic configurations,
+        // and no need to update their configurations with the old values.
```
Comment (seanyen) [inline roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: I am little bit confused by this. Mind to elaborate more?

File: roscpp_azure_iothub/test/testiothub.launch [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<launch>
+
```
Comment (seanyen) [inline roscpp_azure_iothub/test/testiothub.launch:2]: You can define something like:

```xml
<arg name="iot_hub_connection_string" default="abcd"/>
```

And refer it later in the `<param>`:

```xml
<param name="iot_hub_connection_string" value="$(arg iot_hub_connection_string)" />
```

And then user can replace with the string on the fly when running it. For example, `roslaunch roscpp_azure_iothub testiothub.launch iot_hub_connection_string:=<connection_string>`.

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp [hunk 6]
```diff
@@ -204,9 +211,10 @@ void sendMsgToAzureIoTHub(const char* msg, IOTHUB_DEVICE_CLIENT_HANDLE deviceHan
     (void)IoTHubMessage_SetContentTypeSystemProperty(message_handle, "application%2fjson");
     (void)IoTHubMessage_SetContentEncodingSystemProperty(message_handle, "utf-8");
 
-    // Add custom properties to message
-    (void)IoTHubMessage_SetProperty(message_handle, "property_key", "property_value");
-    printf("\r\nSending message %d to IoTHub\r\nMessage: %s\r\n", (int)(messagecount + 1), msg);
+    // Add custom properties to message 
+    //(void)IoTHubMessage_SetProperty(message_handle, "property_key", "property_value");
```
Comment (seanyen) [inline roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: Maybe just remove this line?

**Review by seanyen — APPROVED** [review]
In general, LGTM.

File: roscpp_azure_iothub/test/testiothub.launch (hunk 5 above)
Comment (JiaBaoxi) [reply roscpp_azure_iothub/test/testiothub.launch:2]: Sure. Code updated.

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp (hunk 6 above)
Comment (JiaBaoxi) [reply roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: Sure. Code updated.

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp (hunk 4 above)
Comment (JiaBaoxi) [reply roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: Thanks for the comments. Code updated to remove the global variable to make the logic clear.

File: roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp (hunk 3 above)
Comment (JiaBaoxi) [reply roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp]: Sure. Code updated.

**Pull request merged**